File: web/e-platform-web/src/services/CourseService.js
const API_URL = process.env.REACT_APP_BFF_API_URL || 'https://tunsiaed.onrender.com';

class CourseService {
  /**
   * Fetches all available courses
   * @returns {Promise<Array>} List of all courses
   */
  static async getAllCourses() {
    const res = await fetch(`${API_URL}/api/v1/course`);
    if (!res.ok) throw new Error('Failed to fetch courses');
    return res.json();
  }

  /**
   * Fetches a specific course by ID
   * @param {string} courseId - The ID of the course
   * @returns {Promise<Object>} Course data
   */
  static async getCourseById(courseId) {
    const res = await fetch(`${API_URL}/api/v1/course/${courseId}`);
    if (!res.ok) throw new Error('Failed to fetch course');
    return res.json();
  }

  /**
   * Fetches courses by category
   * @param {string} category - The category name
   * @returns {Promise<Array>} List of courses in the category
   */
  static async getCoursesByCategory(category) {
    const res = await fetch(`${API_URL}/api/v1/course/category/${category}`);
    if (!res.ok) throw new Error('Failed to fetch courses by category');
    return res.json();
  }

  /**
   * Fetches system courses (platform-seeded courses)
   * @returns {Promise<Array>} List of system courses
   */
  static async getSystemCourses() {
    const res = await fetch(`${API_URL}/api/v1/course/system`);
    if (!res.ok) throw new Error('Failed to fetch system courses');
    return res.json();
  }

  /**
   * Fetches all unique course categories
   * @returns {Promise<Array>} List of category names
   */
  static async getAllCategories() {
    const res = await fetch(`${API_URL}/api/v1/course/categories`);
    if (!res.ok) throw new Error('Failed to fetch categories');
    return res.json();
  }

  /**
   * Creates a new course
   * @param {Object} courseData - Course creation data
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Created course data
   */
  static async createCourse(courseData, token) {
    const res = await fetch(`${API_URL}/api/v1/course`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(courseData),
    });
    if (!res.ok) throw new Error('Failed to create course');
    return res.json();
  }

  /**
   * Updates an existing course
   * @param {string} courseId - The ID of the course to update
   * @param {Object} courseData - Updated course data
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Updated course data
   */
  static async updateCourse(courseId, courseData, token) {
    const res = await fetch(`${API_URL}/api/v1/course/${courseId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(courseData),
    });
    if (!res.ok) throw new Error('Failed to update course');
    return res.json();
  }

  /**
   * Deletes a course
   * @param {string} courseId - The ID of the course to delete
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Deletion confirmation
   */
  static async deleteCourse(courseId, token) {
    const res = await fetch(`${API_URL}/api/v1/course/${courseId}`, {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });
    if (!res.ok) throw new Error('Failed to delete course');
    return res.json();
  }

  /**
   * Fetches courses created by the authenticated instructor
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of instructor's courses
   */
  static async getInstructorCourses(token) {
    const res = await fetch(`${API_URL}/api/v1/course/instructor/my-courses`, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });
    if (!res.ok) throw new Error('Failed to fetch instructor courses');
    return res.json();
  }
}

export default CourseService;

File: mobile/e_platform_mobile/lib/Service/course/course_service.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_endpoints.dart';

/// Course Service
/// Handles all course-related API calls
/// Follows exact web architecture
class CourseService {
  final ApiClient _apiClient = ApiClient();

  /// Fetches all available courses
  /// @returns List of all courses
  Future<List<Map<String, dynamic>>> getAllCourses() async {
    final response = await _apiClient.get(ApiEndpoints.courses);
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Fetches a specific course by ID
  /// @param courseId - The ID of the course
  /// @returns Course data
  Future<Map<String, dynamic>> getCourseById(String courseId) async {
    final response = await _apiClient.get(
      ApiEndpoints.courseById(courseId),
    );
    return response.data;
  }

  /// Fetches courses by category
  /// @param category - The category name
  /// @returns List of courses in the category
  Future<List<Map<String, dynamic>>> getCoursesByCategory(String category) async {
    final response = await _apiClient.get(
      '${ApiEndpoints.courses}/category/$category',
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Fetches system courses (platform-seeded courses)
  /// @returns List of system courses
  Future<List<Map<String, dynamic>>> getSystemCourses() async {
    final response = await _apiClient.get(
      '${ApiEndpoints.courses}/system',
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Fetches all unique course categories
  /// @returns List of category names
  Future<List<String>> getAllCategories() async {
    final response = await _apiClient.get(
      '${ApiEndpoints.courses}/categories',
    );
    return List<String>.from(response.data);
  }

  /// Creates a new course
  /// @param courseData - Course creation data
  /// @returns Created course data
  Future<Map<String, dynamic>> createCourse(Map<String, dynamic> courseData) async {
    final response = await _apiClient.post(
      ApiEndpoints.courses,
      data: courseData,
    );
    return response.data;
  }

  /// Updates an existing course
  /// @param courseId - The ID of the course to update
  /// @param courseData - Updated course data
  /// @returns Updated course data
  Future<Map<String, dynamic>> updateCourse(
    String courseId,
    Map<String, dynamic> courseData,
  ) async {
    final response = await _apiClient.put(
      ApiEndpoints.courseById(courseId),
      data: courseData,
    );
    return response.data;
  }

  /// Deletes a course
  /// @param courseId - The ID of the course to delete
  /// @returns Deletion confirmation
  Future<void> deleteCourse(String courseId) async {
    await _apiClient.delete(ApiEndpoints.courseById(courseId));
  }
}
